added support for cotracker for point tracking

parikshit14 opened the pull request:
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/neuroinformatics-unit/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
First step to solve the issue #54 

**What does this PR do?**
 - created base tracker class
 - added an example folder with python script to run cotracker
 - added submodule cotracker to get required functionalities from the official repo

## References

Please reference any existing issues/PRs that relate to this PR.
 - #54 
## How has this PR been tested?

Please explain how any new code has been tested, and how you have ensured that no existing functionality has changed.

## Is this a breaking change?

If this PR breaks any existing functionality, please explain how and why.

## Does this PR require an update to the documentation?

If any features have changed, or have been added. Please explain how the
documentation has been updated.

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

## TODO:
- add unit tests
- update documentation

parikshit14 commented on ethology/trackers/tracker.py:
```diff
@@ -0,0 +1,98 @@
+
+from .cotracker.cotracker.utils.visualizer import (
+    Visualizer,
+    read_video_from_path,
```
Using `read_video_from_path` api from **cotracker** repository

I checked the source code, they use `imageio` with `numpy` to read the video file, which I think is similar to that in movement (which uses `sleap_io` that uses `imageio` as well), so maybe reducing a step

sfmig replied:
sorry what do you mean with reducing a step?

parikshit14 replied:
By reducing a step, I mean reducing an import step, directly using from cotracker to read the video, instead of using movement's way of reading, which also has the same, but a deeper hierarchy. Let me know which is the preferred way to read the video.

parikshit14 commented on ethology/trackers/tracker.py:
```diff
@@ -0,0 +1,98 @@
+                filename=f'{self.tracker}_{video_source.split("/")[-1].split(".")[0]}',
+            )
+
+        return pred_tracks
```
returns a tensor containing tracks of each query point across the frames. so [batch, frame, query_points, X, Y]. Example: 50 frames video with 3 query points will give a tensor of shape [1, 50, 4, 2]

sfmig replied:
Could we include this in the docstring instead?

parikshit14 replied:
Done

sfmig replied:
Feel free to click "Resolve conversation" at the bottom of each conversation thread to mark it as "resolved" if all issues have been addressed. It makes it a bit easier for the reviewer when having a second look. Thanks!

parikshit14 replied:
Hi, sorry for that. I thought you would want to see where you left comments and whether they were addressed.

sfmig replied:
no worries!

parikshit14 commented on examples/tracking_using_cotracker.py:
```diff
@@ -0,0 +1,16 @@
+
+video_source = "../ethology/trackers/cotracker/assets/apple.mp4"
+
+query_points = [
```
query_points have to be of size Q,3 where Q is the number of query points

parikshit14 replied:
Done

parikshit14 commented on ethology/trackers/tracker.py:
```diff
@@ -0,0 +1,98 @@
+
+        self.tracker = tracker_type
+
+    def track(
```
Tried to keep the structure similar to how the function call in ultralytics looks like

parikshit14 commented:
Added unit tests and can add more unit tests based on some feedback

sfmig commented on ethology/trackers/tracker.py:
```diff
@@ -0,0 +1,98 @@
+
+        if save_results:
+            vis = Visualizer(
+                save_dir="../ethology/video",
```
I think it would make sense if the path to the output video is an argument to the function, with a sensible default. I would expect most users would want to place the output videos in specific directories

sfmig commented on tests/test_unit/test_trackers/test_tracker.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@pytest.mark.parametrize(
+    "tracker_type, expected",
```
We can remove the `expected` variable and simply use `tracker_type` to compare

sfmig commented on tests/test_unit/test_trackers/test_tracker.py:
```diff
@@ -0,0 +1,41 @@
+    ],
+)
+def test_init(tracker_type, expected):
+    if isinstance(expected, str):
```
maybe insead: `if tracker_type is in list_supported_tap_models`

sfmig commented on tests/test_unit/test_trackers/test_tracker.py:
Here we are missing a test that checks things work as expected without errors. 

You could do this by formatting your current example as a test. You can maybe have a look at mocking for tests (see an example in the codebase [here](https://github.com/neuroinformatics-unit/ethology/blob/4a9eea668429cc74d86b34e6496fbc98e90f66a2/tests/test_unit/test_annotations/test_load_bboxes.py#L198)).

sfmig commented on examples/tracking_using_cotracker.py:
This example would have been very nice for the PR description, so that the reviewer can quickly see and check how your functionality would be used.

The examples folder here is aimed to support a gallery of examples [like in `movement`](https://github.com/neuroinformatics-unit/movement/tree/main/examples) - these are included in the [documentation](https://movement.neuroinformatics.dev/examples/index.html). We currently don't have the infrastructure for this, so we can omit this from this PR for now.

sfmig commented on ethology/trackers/tracker.py:
```diff
@@ -0,0 +1,98 @@
+        video = torch.from_numpy(video).permute(0, 3, 1, 2)[None].float()
+
+        # load model
+        if self.tracker == "cotracker":
```
We could also add support for a user supplying a checkpoint, using the `CoTrackerPredictor` instead

parikshit14 commented:
Hi @sfmig, I have made the changes requested.

- Updated the structure as mentioned.
- Added cotracker as dependency instead of submodule.
- Added another test that shows proper working, using a mock video as suggested.
- removed the example and transfered the gist of it as a test case.

TODO:
> Ideally, for interoperability with our sister project `movement`, it would return the trajectories as a `movement` dataset. This can be done for example in a `convert_to_movement_dataset` method in this class.
> 
> This may be out of scope for this PR, but if you want to have a go it could be cool.
- I'll work on returing the trajectory as movements_dataset, let me know if you want it to be addressed within this PR or another one. I think it can be done here.


Let me know what you think. Thanks for the references, they were precise and helped speed up the modification.

sfmig commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+        query_tensor = torch.tensor(query_points, dtype=torch.float32)
+
+        # set device
+        device = (
```
Could we include `mps` if available, before considering `cpu`?

parikshit14 replied:
also, when trying to run a larger video. I seem to run out of GPU memory (8 GB). Should the device be taken from user or based on the system itself? current priority cuda > mps > cpu

sfmig replied:
We could allow the user to set it (via config?), but if not set you could have a default priority (cuda > mps > cpu)

sfmig commented on tests/test_unit/test_tap_models/test_tap_models.py:
```diff
@@ -0,0 +1,71 @@
+        ds = tracker.track(
+            video_path="fake_video_path.mp4", query_points=query_points
+        )
+        assert ds.position.shape == (50, 2, 1, 4)
```
For further clarity we can define these as variables and/or derive them from the input data. For example we can have:
```python
n_frames = 50
n_individuals = len(query_points)
n_keypoints = 1
```
and then use them in the assertions

parikshit14 commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+        """
+        # pred_tracks.shape =  [batch, frames, querypoints, 2d points]
+        # goes into positional array with shape
+        # [frames, 2d points, 1 keypoint per query point, no. of query points]
```
Yes, you are correct. Even I was confused looking at the code after a few days, I should have documented it.



> It's a good solution to start with, but it would be very nice to think of a way to make this more flexible. For example, a user may want to specify different names for the individuals or for the keypoint. Or they may want to provide query points that represent _n_ individuals each with _m_ keypoints.

I'll think of a way...

parikshit14 replied:
Have made changes to incorporate multiple keypoints per individual.

parikshit14 commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+        video = video.to(device)
+        query_tensor = query_tensor.to(device)
+
+        pred_tracks, pred_visibility = self.model(
```
Since there is no loop, it seems tqdm will just show 0 or 100, not the progress. Maybe something in cotracker 🤔

parikshit14 replied:
One more question, do you want me to add some print statements after a process completes, like once the video is processed through the model maybe have a print statement, similarly after saving the video? It might get a little verbose as this repo adds further functionalities, but it may help with the status of execution of the processes.

sfmig replied:
Yeah maybe some logging (with print statements for now) already helps. We can polish that in the future.

Again somewhat out of scope for this PR, so feel free to ignore, but we could consider using [loguru](https://github.com/Delgan/loguru) for logging rather than using print statements. It provides some more flexibility

parikshit14 commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+        video_path: str,
+        query_points: list[list],
+        save_dir: str = "processed_videos",
+        save_results: bool = False,
```
I thought of that, but since you asked last time if we could give some sensible default for the save dir, there wasn't a way out.

sfmig replied:
Yeah sometimes there is a bit of back and forth. I'd say let's try to keep it as simple as possible to start with, and reduce the number of input arguments? What do you think?

parikshit14 replied:
This makes sense, changed!

parikshit14 commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+                f'{self.tracker}_{video_path.split("/")[-1].split(".")[0]}'
+            )
+
+            vis = Visualizer(
```
I have a cool idea about it.. Maybe we can structure this like how mmdetection (https://github.com/open-mmlab/mmdetection) has done it. Most people are aware of detection/segmentation, might know how the flow is, and it will help us manage better.

So running the file might look like 
```
python tap_models/track_any_point.py --config configs/tap_models/cotracker.yaml
```

In similar way when adding trackers we can use something like `configs/trackers/bytetrack.yaml` , etc.
Let me know what you think....

sfmig replied:
Yeah, something like that could work nicely!

> Most people are aware of detection/segmentation, might know how the flow is, and it will help us manage better.

Good point! I think this is reasonable to assume.

Ideally, we would maintain a programmatic way to run inference with the model too (so not just a command-line interface), and probably somewhat prioritise that over the CLI access. So that may look like, for example, a `BaseTrackAnyPoint` class that would also accept receiving a config file as input upon construction. And we could have default config files with reasonable values that we ship with the package. Does that make sense?

If you want, we can focus on the API/programmatic access for this PR - just to keep it simple for now and keep this PR reasonably sized. Once we have that, adding a CLI on top of that should be straightforward.

There are some disadvantages that we have noted in the team when developing CLI tools too. They can be hard to maintain and use when you have lots of options (some of which could be mutually exclusive), and we find often most experienced users end up falling back on the Python API anyways. A CLI in our experience seems to work better with tools that are a bit more narrow in their functionality. But you are right they are a very common way to interact with with ML models for training, running inference etc. so we may have to reasses this going forward. I think the lesson learnt is in a way: let's try not to introduce unnecessary complexity early on, as that may be hard to undo.

parikshit14 commented on ethology/tap_models/track_any_point.py:
```diff
@@ -0,0 +1,163 @@
+        )
+        return ds
+
+    def track(
```
I have pushed some commits to restructure the class; some functions remain in the class, while some have moved out. Let me know your views.

parikshit14 commented:
I think we can conclude this PR other changes I have in mind are
- changing the `save_video` method to plot trajectories on our own instead of relying on Cotracker's visualizer since we now support multi-keypoint support as well
- Customize the configurations from a **config.yaml** file
- Customize FPS (however, it would be better if I could get some idea about the purpose behind doing this customization)

but they feel out of the scope of this one.

sfmig reviewed (commented):
Hi @parikshit14, thanks for the update.

I noticed the latest CI checks are failing - would you mind having a look?

I also had a quick look at the latest state and left some in-line comments.

I think you make a good point re visualisation in your last comment, and would actually like to know your views on the following suggestion:

I think we could do away with the visualisation bits actually. What I have been doing lately as a user in a similar setup is exporting the predictions as a [movement pose dataset](https://movement.neuroinformatics.dev/user_guide/movement_dataset.html), which I can then export in a few formats and importantly, I can load in napari. This removes a lot of the faff involved in exporting custom videos and could make our code (and in particular this PR) much simpler. Would you like to have a go at this? 

If so, I would recommend having a look at [this guide](https://movement.neuroinformatics.dev/user_guide/gui.html) first to get a feel for it. In movement, I am currently working on exporting the bboxes dataset to VIA format, which is also a format importable in napari, so that should help in this too.

Let me know thoughts. If you feel you don't have capacity to continue this work further, I understand and I would be happy to take it from here (your contributions will still be logged and acknowledged of course). I realise it may not feel the best to keep iterating on this PR, but unfortunately I think that is the nature of a package in such early stages, in which we are still figuring out how to best implement things and don't have much codebase to use as reference.

parikshit14 commented:
> I think we could do away with the visualisation bits actually. What I have been doing lately as a user in a similar setup is exporting the predictions as a [movement pose dataset](https://movement.neuroinformatics.dev/user_guide/movement_dataset.html), which I can then export in a few formats and importantly, I can load in napari. This removes a lot of the faff involved in exporting custom videos and could make our code (and in particular this PR) much simpler. Would you like to have a go at this?

> If so, I would recommend having a look at [this guide](https://movement.neuroinformatics.dev/user_guide/gui.html) first to get a feel for it. In movement, I am currently working on exporting the bboxes dataset to VIA format, which is also a format importable in napari, so that should help in this too.

I am not sure if you had a look at my recent commits, but they address this exactly using the function `convert_to_movement_dataset` (I haven't checked with napari as of now).

> Let me know thoughts. If you feel you don't have capacity to continue this work further, I understand and I would be happy to take it from here (your contributions will still be logged and acknowledged of course).

I'm disappointed that I made you feel this way, since even after the contribution period ended, I still made regular commits and waited for your reviews.

sfmig commented:
Hi @parikshit

Some answers to your points below.

> I am not sure if you had a look at my recent commits, but they address this exactly using the function `convert_to_movement_dataset` (I haven't checked with napari as of now).

Yes, I remember this came from a PR review comment and I am aware this is in the code. I should have been more clear about my suggestion: what I meant is that we can remove any video export utilities from the PR (such as `save_video`) and rely on napari support for visualisation instead (via `movement` for now). It would mean something like replacing `save_video` by a "save tracked output" function, that wraps a `movement` export function (maybe [as a SLEAP file](https://movement.neuroinformatics.dev/api/movement.io.save_poses.to_sleap_analysis_file.html#movement.io.save_poses.to_sleap_analysis_file) and/or as a [DLC file](https://movement.neuroinformatics.dev/api/movement.io.save_poses.to_dlc_file.html)). As long as we export the tracks in a `movement`-compatible format that is loadable in napari (which we are), it should be good.

I would address that as a next step, along with the pending CI checks that are currently failing. Another bit for improvement imo would be the out-of-memory error handling, trying to downscale the input image for example, but that requires a bit more thinking and we could address that in a future separate PR.

> I'm disappointed that I made you feel this way, since even after the contribution period ended, I still made regular commits and waited for your reviews.

I'm sorry to hear that you are disappointed by my previous message. I wonder if it was misinterpreted. Just to clarify, with "capacity" I meant "time capacity", referring to the total number of hours a person can potentially work on something over a given period. I didn't mean it as a synonym for "ability". It is a slightly business-y term that we use quite a bit in the team but I realise it may be uncommon in other contexts.

I hope this helps, I want to be clear that I appreciate your efforts and interest in pushing this forward and have enjoyed working collaboratively on this.

parikshit14 commented:
Hi @sfmig, sorry for the delay. I'll work on it and push commits soon.


> I'm sorry to hear that you are disappointed by my previous message. I wonder if it was misinterpreted. Just to clarify, with "capacity" I meant "time capacity", referring to the total number of hours a person can potentially work on something over a given period. I didn't mean it as a synonym for "ability". It is a slightly business-y term that we use quite a bit in the team but I realise it may be uncommon in other contexts.

What I meant was that I was disappointed that I gave you any hint that made you feel I wouldn't contribute to this PR anymore.
